### PR TITLE
Using a class variable instead of exception for a performance boost

### PR DESCRIPTION
**Pull request by ray-delossantos**

### CODE REVIEW

File: src/toro.cr [hunk 1]
```diff
@@ -53,8 +53,10 @@ module Toro
     getter context : HTTP::Server::Context
 
     @inbox = Hash(Symbol, String).new
+    @halt = false
 
     def initialize(@context)
+      @halt = false
```
Comment (ray-delossantos) [inline src/toro.cr:59]: @soveran , with this change you'll get a 30% increment in the speed of the router (benchmarked with wrk in a linux machine)

**Comment by soveran** [comment]
This is amazing, I'm trying it locally and the speedup is huge!

I discovered one drawback that we should consider, which is the fact that all the matchers at the same level and above the one that halted are still evaluated (not the blocks, but the matchers themselves). I'll illustrate the case with an example:

``` crystal
class App < Toro::Router
  def routes
    on "foo" do
      text "here"
    end

    text "path: #{path.prev}"

    on "bar" do
      text "not here"
    end

    text "path: #{path.prev}"
  end
end

Toro.drive(App, "GET", "/foo/bar").body
#=> "here\npath: /foo\npath: /foo/bar\n"
```

It means the call to `on "bar"` didn't yield to the block, yet it tried to match the path segment (and succeeded). I don't think it could be problematic in this case, but now let's see another scenario:

``` crystal
class App < Toro::Router
  def routes
    on notify_admins do
      ...
    end

    on notify_customer do
      ...
    end
  end
end
```

Let's say `notify_admins` and `notify_customer` return booleans after scheduling an email delivery. Let's say `notify_admins` returns `true`. The problem here is that even if the block after `notify_customer` is not evaluated, the call to `notify_customer` is executed anyway, which is not what we want.

The other minor drawback would be that if we have many matchers at the same level, all those methods calls can start to add up (even if they don't yield to the blocks when the application is halted). But the speedup of using an instance variable instead of an exception is so big that I think the change will be worth doing anyway.

So right now I want to think about some ways to prevent the first problem. It makes me sad to think that if we can't prevent it, we may not be able to use this approach, but I have some ideas to explore and I'm hopeful at this point. In any case, it's great that you read the code and were able to spot all these details, I'm extremely glad you got interested in this project :-)

Again thanks, and I'll work on this a bit more to see if there's a workaround.

**Comment by soveran** [comment]
I may have found a way :-)

Still some tests to do, stay tuned!

**Comment by ray-delossantos** [comment]
I'm working on a helper method for json responses, after you rebase this pull request i'll love to show some of the changes for workng with json. I compared the performance of this library and is around 3-4x faster than kemal, that's why i'm very interested in the library and for me it feels easier to understand. 

I was working on an internal web framework, and after i saw Toro i'm already moving everything to this new project. 

Later i'l try to write some code examples with a JWT authentication middleware.

**Comment by soveran** [comment]
Ah, sorry I didn't reply about the JSON question in the other thread. Adding JSON support is indeed very appealing to me, so good idea about exploring that feature! I'll let you know when this is ready, I hope it will be in the next couple hours. Good work!

File: src/toro.cr [hunk 2]
```diff
@@ -134,6 +147,16 @@ module Toro
 
     abstract def routes
 
+    def json(response : T, status_code : Int? = nil)
```
Comment (ray-delossantos) [inline src/toro.cr:150]: This is the new json method for outputing a json from a generic object.

File: src/toro.cr [hunk 3]
```diff
@@ -89,28 +92,38 @@ module Toro
...
-      yield
+      if !@halt
+        yield
+      end
       halt
     end
 
     def on(cond : Bool)
-      default { yield } if cond
+      if !@halt
```
Comment (ray-delossantos) [inline src/toro.cr:102]: Fixed the problem with the evaluation of the route segment.

File: spec/toro_spec.cr [hunk 4]
```diff
@@ -292,3 +292,25 @@ describe "basic auth" do
     assert_equal "user:1\n", response.body
   end
 end
+
```
Comment (ray-delossantos) [inline spec/toro_spec.cr:295]: Added spec for the .json helper method.

File: src/toro.cr (hunk 2 above)
Comment (soveran) [inline src/toro.cr:150]: Can you put this commit in a separate pull request? That way I can merge only the halt changes.

Comment (ray-delossantos) [inline src/toro.cr:150]: I really don't know how to do that in git. But i'll keep investigating. Sorry for the mess :).

**Comment by soveran** [comment]
I just pushed the `unstable` branch, where I changed part of the API to use macros instead of methods. With that change, we can now return immediately once a matcher succeeds and after it yields. It means we have the best of both worlds: we don't use exceptions, and we skip all the remaining matchers, which was the concern I had with the `@halt` approach. If you like this approach, I will release a new version and I will add comments on how you helped speed it up. If it weren't for your suggestion to use `@halt`, I think we would have never seen this change applied.

Also, if you like this change, you can apply the JSON change to this branch and we can start working on it. Waiting for your input, thanks a lot for the idea!

**Comment by ray-delossantos** [comment]
I checked the unstable branch, excellent and clean solution for the problem. I'll rebase after your changes and make a new pull request with the json helper, specs and documentation.

**Comment by soveran** [comment]
Excellent! I'll merge this to master, but it's a good idea if your pull request goes to the `unstable` branch. I think that will make it easier for us to collaborate on this project. Again, thanks a lot for what you are doing!

**Comment by ray-delossantos** [comment]
I'll create a new branch for the json feature in my repo, and a pull request against unstable.

**Comment by soveran** [comment]
@raydf Perfect, I will close this pull request now and we continue in the other one once it's ready.

**Pull request closed without merging**